---
title: "HCC Case 1"
output:
  html_document:
    theme: united
    toc: yes
  pdf_document:
    toc: yes
---


The only variable which needs to be changed is "base.folder". If the user's file structure matches the example, other variables are generated automatically.

```{r setup, include=TRUE, echo=TRUE, message=FALSE, warning=FALSE}
start.time <- Sys.time()
# knitr::opts_knit$set(root.dir = '../..')
base.folder <- '~/octad_desktop/' # should be set to installation parent folder
setwd(base.folder)
library(dplyr)
library(ggplot2)
library("RUVSeq")
library("GSVA")
library(compiler)
library(data.table)
library(devtools)
library(limma)
library(rhdf5)
# library(doParallel)


enableJIT(3)


outputFolder = paste0(base.folder,'/results/case1_hcc_aenormal/')
if (!dir.exists(outputFolder)) {
  dir.create(outputFolder)
}
dataFolder = paste0(base.folder,'data/')
CodeFolder = paste0(base.folder,'code/')

load(paste0(dataFolder,'metadata.RData')) # loads 'ensemble_info', 'merged_gene_info', 'breastpam50', 'tsne', and 'phenoDF'

source(paste0(CodeFolder,'core_functions.R'))

# avoiding logging errors
tsne.time    <- "Not run"
GEA.time     <- "Not run"
VDH.time     <- "Not run"
enrich.time  <- "Not run"
ranking.time <- "Not run"
```

Here, you will normalize samples for batch effects with package "RUVSeq". You can set it off by changing set parameters script.   
Set parameters. Below are default settings. 
You can set normalize_samples = F to make it run faster at risk of batch effects.  
```{r parameters}
setwd(base.folder)
# DE params:

normalize_samples = T
k = 1
n_topGenes = 10000
#this demo sample set does not have too much genes. The usual default for n_topGenes = 10000. Set this higher if you think there are more significant genes
DE_method = 'edgeR' # Other choice is 'limma' for DE_method.


# disease signature params:
log2fchange <- 2 # the cutoff for "significant gene impact". Default = 1, so any DE gene with log2foldchange < 1 will be omitted. Improved results with 2.
padjusted <- 0.001 # the cutoff for "significant genes". Default = 0.001, so any DE genes with padj > 0.001 will be omitted.


# sRGES params:
# Only mess with "max_gene_size", which is how many up-regulated genes and down-regulated genes are allowed. ("100" = 100up + 100down = 200 total)
landmark = 1
choose_fda_drugs = F
max_gene_size = 100
weight_cell_line = F

# logging parameters
write('x', file = paste0(outputFolder,"parameters.txt"))
fileConn<-file(paste0(outputFolder,"parameters.txt"))
writeLines(c("Normalization Parameters:","normalize_samples",normalize_samples,"","k",k,"","n_topGenes",n_topGenes,"","DE_method",DE_method,"","--------------------------","","Disease Signature Parameters:","log2fchange",log2fchange,"","padjusted",padjusted,"","--------------------------","","sRGES Parameters:","landmark",landmark,"","choose_fda_drugs",choose_fda_drugs,"","max_gene_size",max_gene_size,"","weight_cell_line",weight_cell_line), fileConn)
close(fileConn)

```

Have the following files in your dataFolder:

* metadata.RData  
* octad.h5 [expression data]    
* cmpd_sets_chembl_targets.RData  
* cmpd_sets_mesh.RData  
* cmpd_sets_meshes.RData  
* cmpd_sets_sea_targets.RData  
* cmpd_sets_ChemCluster.RData
* encoderDF_AEmodel1.RData  
* lincs_sig_info.csv  
* lincs_signatures_cmpd_landmark_symbol.RData  
* CCLE_OCTAD.RData  
* repurposing_drugs_20170327.csv  
* metadata.RData  

```{r dataFolder, results='hide'}
setwd(base.folder)
dir(dataFolder)
#check to make sure you have these files
```
***

Have the following scripts in your CodeFolder:  
  
* [your workflow].Rmd (this Rmd is an example)  
* core_functions.R  
```{r CodeFolder}
setwd(base.folder)
dir(CodeFolder)
#check to make sure you have these files
```
***

##Select Cases##    
```{r case.load1}
setwd(base.folder)
case.load.start <- Sys.time()

#lumA_over50    <- breastpam50 %>% filter(LumA > 0.5)
phenoDF_case   <- phenoDF %>% filter(cancer == 'Liver Hepatocellular Carcinoma', sample.type == 'primary', data.source == 'TCGA')
case_id = phenoDF_case$sample.id
```
***  

##Select Reference Control Samples##  
***
```{r case.load3}
setwd(base.folder)
setwd(base.folder)
normal_id = (phenoDF %>% filter(data.source == 'GTEX'))$sample.id


```

```{r}
load(paste0(dataFolder,'encoderDF_AEmodel1.RData'))
normal_counts = EncoderDF[,normal_id]
case_counts = EncoderDF[,case_id]

(row.names(case_counts) == row.names(normal_counts)) %>% sum() # should be 64
#quick check to see that all genes match up before combining  

dz_expr = cbind(normal_counts, case_counts)

rm(normal_counts, case_counts) # free up some memory

control_id <- computeRefTissue(case_id = case_id, #computes ref tissue given case_id, normal_id chr vectors
                               normal_id = normal_id,
                               expSet = dz_expr, #expSet is a counts dataframe that contains both case and normal counts. 
                               control_size = 50) 
write(control_id, file = paste0(outputFolder,'control_ids.txt'))
#this function will also generate a csv called case_normal_median_cor.csv that includes the median correlation of each normal sample.

normal_median_cor = read.csv(paste0(outputFolder,'case_normal_median_cor.csv'),stringsAsFactors = F)
normal_median_cor <- normal_median_cor %>% arrange(desc(cor))

normal_median_cor <-left_join(normal_median_cor, 
  phenoDF %>% select(sample.id, biopsy.site), 
  by = "sample.id")

head(normal_median_cor)

```

***

```{r case.load6}

setwd(base.folder)
rhdf5_file <- paste0(dataFolder, "octad.h5")
transcripts = as.character(h5read(rhdf5_file, "meta/transcripts"))
samples = as.character(h5read(rhdf5_file, "meta/samples"))
case_counts = h5read(rhdf5_file, "data/count", index=list(1:length(transcripts), which(samples %in% case_id)))
colnames(case_counts) = samples[samples %in% case_id]
rownames(case_counts) = transcripts
case_id = samples[samples %in% case_id]
normal_counts = h5read(rhdf5_file, "data/count", index=list(1:length(transcripts), which(samples %in% control_id)))
colnames(normal_counts) = samples[samples %in% control_id]
rownames(normal_counts) = transcripts
control_id = samples[samples %in% control_id]
H5close()

dz_expr = cbind(normal_counts, case_counts)
rm(normal_counts) # free up some memory
rm(case_counts)

case.load.end <- Sys.time()

case.load.time <- case.load.end - case.load.start
case.load.time
```

##Visualize on Cancer Map##
Optional case-control visualization.
```{r cancermap, fig.height=10, fig.align= "center"}
#optional - duration: approximately 3 seconds
setwd(base.folder)

tsne.time.start <- Sys.time()

#make a new column called type to state case, control, or others
tsne$type <- "others"
tsne$type[tsne$sample.id %in% case_id] <- "case"
tsne$type[tsne$sample.id %in% control_id] <- "control"

#plot 
(p2 <- ggplot(tsne, aes(X, Y, color = type)) + geom_point(alpha = 0.4)+
    labs(title = paste ('TNSE PLOT'), x= 'TSNE Dim1', y='TSNE Dim2', caption="OCTAD")+
    theme_bw())

ggsave("case_control_map.pdf", path=outputFolder)

tsne.time.stop <- Sys.time()
tsne.time <- tsne.time.stop - tsne.time.start
tsne.time
```

##Run Differential Expressions##  

Variables needed:  

* case_id : samples IDs of case tissues in a character vector
* control_id : samples IDs of control tissues in a character vector
* dz_expr : dataframe that contains both case_id and control_id in its column names, 
* outputFolder : folder to place output files (must have this set above)

***

DESeq takes too long so I use a demo set here
```{r DE2, message=FALSE, warning=FALSE}
setwd(base.folder)

DE.start <- Sys.time()

# important that the below res is generated LAST.

res = diffExp(case_id = case_id,
        control_id = control_id,
        expSet = dz_expr,
        normalize_samples = T,
        DE_method='edgeR'
        )
DE.end <- Sys.time() - DE.start
print(DE.end)
res_geneinfo <- left_join(res, merged_gene_info, by=c('identifier'='ensembl'))
```

***

Compare edgeR res to LIHC_sig from nature paper
```{r}
res_geneinfo = read.csv(file='~/octad_desktop/results/case1_hcc_aenormal/res_geneinfo.csv',stringsAsFactors = F)
LIHCsig = read.csv(file='../results/LIHC_sig.csv',stringsAsFactors = F)
LIHCval = inner_join(LIHCsig,res_geneinfo %>% select(GeneID,log2FoldChange,padj)
,by='GeneID',suffix = c('.edgeR','.natComm'))
#LIHCval = LIHCval %>% filter(padj.y<0.1)
plot(LIHCval$log2FoldChange.edgeR,LIHCval$log2FoldChange.natComm)
cor.test(LIHCval$log2FoldChange.edgeR,LIHCval$log2FoldChange.natComm)
```

Get the disease signature.
```{r dz_signature}
setwd(base.folder)
dz_signature <- res %>% filter(padj < padjusted, abs(log2FoldChange)>log2fchange)
#filter out res df to get dz sigs
#Term res refers to the results table that is generated from the Diff_Exp. R script.
dim(dz_signature)
head(dz_signature)
```
***

Combine the res table with the Ensembl table to generate a differential expression results table with gene information.
```{r res_geneinfo}
setwd(base.folder)
res_geneinfo <- left_join(res, merged_gene_info, by=c('identifier'='ensembl'))


write.csv(res_geneinfo,file=paste0(outputFolder,'res_geneinfo.csv'))
```
***

Generate table for disease signature
```{r dz_signature2}
setwd(base.folder)
dz_signature <- dz_signature %>% left_join(ensembl_info %>% select(gene,ensembl,chrom,strand),
                                           by=c('identifier'='ensembl'))
dim(dz_signature)
head(dz_signature)


#the column name may differ between different metadata but you need Symbol to join with RGES
dz_signature$Symbol <- toupper(dz_signature$gene)
str(dz_signature$Symbol)


write.csv(dz_signature, paste0(outputFolder, "/dz_signature.csv"),row.names = F)
```
***

##Affected Pathways and Potential Drugs for Reversal##  
***
Find potential drugs that may reverse disease signature.  Takes ~10mins to run following chunk. Following scripts will output "all_lincs_score.csv"" and "sRGES.csv".  
File all_lincs_score.csv has the RGES (reversal gene expression score), which is needed for the gene_enrichment_analysis script later.    
File sRGES.csv is the summarized RGES.
```{r sRGES}
setwd(base.folder)
sRGES.start <- Sys.time()
# source(paste0(CodeFolder,'core_functions.R'))
#functions needed to run RGES
# source(paste0(CodeFolder,'runRGES_dz_arrangeMax_compiler.R'))
runsRGES(
  dz_signature = dz_signature,
  choose_fda_drugs = choose_fda_drugs,
  parallel = F,                         # causes issues for some machines.
  max_gene_size = max_gene_size,
  landmark = landmark
  )

sRGES.end <- Sys.time()
sRGES.time <- sRGES.end - sRGES.start
print(sRGES.time)
```

***
Compare sRGES from edgeR (this run) with nature comm
```{r}
sRGES = read.csv('~/octad_desktop/results/case1_hcc_aenormal/sRGES.csv',stringsAsFactors = F)
#sRGES_HEPG2 = read.csv('~/octad_desktop/results/case1_hcc/sRGESHEPG2.csv')
LIHCdrugs = read.csv('../octad_case/dataset/LIHC_rges_ic50_normalized.csv',stringsAsFactors = F)
sRGES$pert_iname = toupper(sRGES$pert_iname)
LIHCdrugs$pert_iname = toupper(LIHCdrugs$pert_iname)
sRGESval = inner_join(sRGES,LIHCdrugs,by='pert_iname',suffix=c('.octad','.natcomm'))

plot(sRGESval$sRGES.octad,sRGESval$sRGES.natcomm)
cor.test(sRGESval$sRGES.octad,sRGESval$sRGES.natcomm)

```

```{r}
setwd(base.folder)
enrich.time.start <- Sys.time()

source(paste0(CodeFolder,'core_functions.R'))
#targets = c('chembl_targets','mesh','sea_targets','ChemCluster') # include as many or few as you prefer. 
targets = c('chembl_targets')


enrichFolder <- paste0(outputFolder,'enrichment_analysis/')
if (!dir.exists(enrichFolder)) {
  dir.create(enrichFolder)
}

sRGES = read.csv(paste0(outputFolder,'/sRGES.csv'),stringsAsFactors = F)
load(paste0(dataFolder,"random_gsea_score.RData"))

for (target_type in targets){
  drug_enrichment(sRGES = sRGES, target_type = target_type)
}


enrich.time.stop <- Sys.time()
enrich.time <- enrich.time.stop - enrich.time.start
enrich.time
```
```{r}
## Find best cell line & Validate with cell line data in silico ##
#optional - Duration: approximately 20 seconds.
setwd(base.folder)
ranking.time.start  <- Sys.time()

toplineDF <- computeCellLine(case_id = case_id,
                           expSet = dz_expr,
                           LINCS_overlaps = T,
                           returnDF = T)
topLineEval(topline = c('HEPG2'),mysRGES = sRGES) 
#if we only use the sRGES_HEPG2 subset result is very poor

ranking.time.stop  <- Sys.time()
ranking.time <- ranking.time.stop - ranking.time.start
ranking.time

```
